### ISSUE TITLE
Android ustream

### ISSUE THREAD

**Issue by unlocKing** [beardypig/plugin.video.streamlink#2]
I keep getting the following errors trying to watch ustream.tv/channel/C68tFucrcTa through Android, plays fine in Windows:

21:07:40.845 T:18446744072379432136   ERROR: Failed to load plugin streann:
21:07:40.850 T:18446744072379432136   ERROR:   File "/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/plugin.video.streamlink/lib/streamlink/plugins/streann.py", line 14, in <module>
                                                from streamlink.utils.crypto import decrypt_openssl
21:07:40.850 T:18446744072379432136   ERROR:   File "/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/plugin.video.streamlink/lib/streamlink/utils/crypto.py", line 4, in <module>
                                                from Crypto.Cipher import AES
21:07:40.850 T:18446744072379432136   ERROR:   File "/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/plugin.video.streamlink/lib/streamlink/utils/Crypto.py", line 4, in <module>
                                                from Crypto.Cipher import AES
21:07:40.850 T:18446744072379432136   ERROR: ImportError: No module named Cipher
21:07:44.185 T:18446744072379432136   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <class 'streamlink.exceptions.PluginError'>
                                            Error Contents: [Errno 2] No such file or directory: 'databases/iso639-3.json'
                                            Traceback (most recent call last):
                                              File "/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/plugin.video.streamlink/addon.py", line 85, in <module>
                                                plugin.run()
                                              File "/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/plugin.video.streamlink/simpleplugin.py", line 779, in run
                                                result = action_callable(params)
                                              File "/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/plugin.video.streamlink/addon.py", line 56, in root
                                                streams = session.streams(url)
                                              File "/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/plugin.video.streamlink/lib/streamlink/session.py", line 424, in streams
                                                return plugin.streams(**params)
                                              File "/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/plugin.video.streamlink/lib/streamlink/plugin/plugin.py", line 246, in streams
                                                raise PluginError(err)
                                            PluginError: [Errno 2] No such file or directory: 'databases/iso639-3.json'
                                            -->End of Python script error report<--

**Comment by beardypig** [comment]
@unlocKing sorry, totally missed this! I'll take a look when I can.

**Comment by unlocKing** [comment]
No problem dude..
 
Would be great if you can get it sorted with the 0.7 release of Streamlink.

Thanks.

**Comment by back-to** [comment]
I had the same issue on **FireTV ARMv7**

---

for 
```
No module named Cipher
```

you could use **Kodi v18** with `script.module.pycryptodome`

https://github.com/xbmc/xbmc/blob/master/addons/script.module.pycryptodome/addon.xml

but you need to change every import from `Crypto` to `Cryptodome`

---

for
```
PluginError: [Errno 2] No such file or directory: 'databases/iso639-3.json'
```

I did not find a way for this when i tried,
it's just to annoying to **DEBUG stuff on Android** without real root or file access

---

I just use `livestreamer` instead of `streamlink`, since you don't need this other scripts there,
for the plugins that I use its enough.

**Comment by unlocKing** [comment]
@back-to I had your streamlink module working as that was using extra files - you have taken it down now from your git.. any chance you and @beardypig can work together on the future development of this?

Cheers!

**Comment by back-to** [comment]
hm weird i had issues like this aswell on Android,

I abort it since i use `livestreamer`

---

the only big differnce was i did split this plugin in two plugins instead of using the lib in the same plugin

https://github.com/beardypig/plugin.video.streamlink/blob/master/addon.py#L6

**Comment by unlocKing** [comment]
@back-to Maybe it was the Kodi version you're using, I'm still using Jarvis as that loads live tv streams faster than Krypton.

Do you have your own livesteamer module?

**Comment by back-to** [comment]
yeah i was on Krypton, i tested v18 today but i had issues there aswell.

Well my livestreamer plugin is like this plugin, but split in two plugins and without the other scripts.

I did not test ustream with livestreamer, not sure if it will work.

**Comment by unlocKing** [comment]
@back-to On Android I'm using SPMC 16.70 - has the https and subtitles fix, works really well.

If you get a chance can you test the ustream for me, would really appreciate it.

Cheers!

**Comment by back-to** [comment]
I solved the Issues with Android

---

```
PluginError: [Errno 2] No such file or directory: 'databases/iso639-3.json'
```

Android has some issues with `from pkg_resources import resource_filename`

https://github.com/neverreply/script.module.pycountry/commit/275fc760e08cb8160d01f97bb55ac89572e799b6

If i get the path from the Addon itself it works.

---

```
ERROR: ImportError: No module named Cipher
```

https://github.com/streamlink/streamlink/blob/master/src/streamlink/utils/crypto.py

must be renamed to something else

https://github.com/neverreply/script.module.streamlink.base/blob/master/resources/lib/streamlink/utils/cryptofix.py

and after this every import must be rename

```py
# from
from streamlink.utils.crypto import ...
# to
from streamlink.utils.cryptofix import ...
```

---

For **ustream** it will close after 30 sec because of

https://github.com/beardypig/plugin.video.streamlink/blob/master/addon.py#L69

it must be **ignored there** to get to

https://github.com/beardypig/plugin.video.streamlink/blob/master/addon.py#L81

i think it worked then not sure.
Tested it on NASA stream but its sometimes black so yeah.

---

I'm **currently testing** to run streamlink as a service with XBMCLocalProxy

it has still some issues like
- like after install/update, it requires a restart 🌊 
- lag at 1080P60FPS, but thats maybe the Androidbox

https://github.com/neverreply/service.streamlink.proxy

---

tl;dr

**Android is stupid**

**Comment by unlocKing** [comment]
Great progress..please keep us updated.

The neverreply account is that yours also?
